homeworks/hw12/src/bank.rs

```diff
@@ -0,0 +1,999 @@
+//! This module contains the implementation of a simple banking system.
+//!
+//! The [`Bank`] struct represents a bank and provides methods for managing accounts
+//! and performing various banking operations such as deposits, withdrawals, and transfers.
+///
+use log::{debug, error, info};
+use rand::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::slice::Iter;
+use thiserror::Error;
+
+type Money = f64;
+pub type Result<T, E = BankError> = std::result::Result<T, E>;
+
+const MONEY_ZERO: Money = 0.0;
+
+type TransactionId = String;
+
+#[derive(Default)]
+pub struct Bank {
+    accounts: HashMap<String, RefCell<Money>>,
+    accounts_history: HashMap<String, Vec<TransactionId>>,
+    history: BTreeMap<TransactionId, Operation>,
+    ulid_generator: ulid::Generator,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Operation {
+    #[allow(unused)]
+    id: String,
+    source_account: String,
+    amount: Money,
+    operation_type: OperationType,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum OperationType {
+    CreateAccount,
+    Deposit,
+    Withdraw,
+    Transfer { target_account: String },
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Account already exists")]
+pub struct AccountDuplicationError {
+    account: String,
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error("Account does not exist")]
+pub struct AccountNotFoundError {
+    account: String,
+}
+
+#[macro_export]
+macro_rules! account_not_found_error {
+    ($account: expr) => {{
+        error!("Account {} does not exist", $account);
+        Err(AccountNotFoundError { account: $account }.into())
+    }};
+}
```

Comment:
Кажется, это можно заменить на функцию. Даже на метод для `BankError`:

```suggestion
impl BankError {
    pub fn account_not_found(account: String) -> Self {
        error!("Account {} does not exist", $account);
        AccountNotFoundError { account: $account }.into()
    }
}
```

Распространенная практика в таких ситуациях